doc: improve fetchers overview, deduplicate readme content, follow doc conventions (#297654)
NixOS/nixpkgs: doc/build-helpers/fetchers.chapter.md

  Nixpkgs fetchers only re-download if the specified hash changes or the store object is not available.
- Built-in fetchers do not use [substituters](https://nixos.org/manual/nix/stable/command-ref/conf-file#conf-substituters).
  Derivations produced by Nixpkgs fetchers will use any configured binary cache transparently.

This significantly reduces the time needed to evaluate code in Nixpkgs, and allows [Hydra](https://nixos.org/hydra) to retain and re-distribute sources used by Nixpkgs in the [public binary cache](https://cache.nixos.org).
For these reasons, the Nix built-in fetchers are not allowed in Nixpkgs code.

The following table summarises the differences:

| Fetchers | Download | Output | Cache | Re-download when |
|-|-|-|-|-|
| `builtins.fetch*` | evaluation time | store path | `/nix/store`, `~/.cache/nix` | `tarball-ttl` expires, cache miss in `~/.cache/nix`, output store object not in local store |
| `pkgs.fetch*` | build time | derivation | `/nix/store`, substituters | output store object not available |

:::{.tip}
`pkgs.fetchFrom*` helpers retrieve _snapshots_ of version-controlled sources, as opposed to the entire version history, which is more efficient.
:::

> 

Author: I'll include `fetchgit` in the note. Regarding mirroring: I'll come back to this and update it with mirroring information as I make my way through documenting all the fetchers, because then I'll see which fetcher supports mirroring and which doesn't, whether there are any differences in support, and so on.

It's likely this tip will go away once all fetchers are better documented, because by then I'll have a better idea of how to organise all this fetcher-agnostic content as well.

5. []{#sec-pkgs-fetchers-updating-source-hashes-fakehash-method} In your package recipe, set the hash to one of

   - `""`
   - `lib.fakeHash`
   - `lib.fakeSha256`
   - `lib.fakeSha512`

   Once you attempt to build, error messages with the calculated hashes will show up, which you then replace in the recipe.

   :::{.warning}
   You must use one of these four fake hashes and not some arbitrarily-chosen hash.
   See [](#sec-pkgs-fetchers-secure-hashes) for details.
   :::

:::{.example #ex-fetchers-update-fod-hash}

# Update source hash with the fake hash method

Consider the following recipe that produces a plain file:

```nix
{ fetchurl }:
fetchurl {
  url = "https://raw.githubusercontent.com/NixOS/nixpkgs/23.05/.version";
  hash = "sha256-ZHl1emidXVojm83LCVrwULpwIzKE/mYwfztVkvpruOM=";
}
```

A common mistake is to update a fetcher parameter, such as `url`, without updating the hash:

```nix
{ fetchurl }:
fetchurl {
  url = "https://raw.githubusercontent.com/NixOS/nixpkgs/23.11/.version";
  hash = "sha256-ZHl1emidXVojm83LCVrwULpwIzKE/mYwfztVkvpruOM=";
}
```

**This will produce the same output as before!**

> To determine the new hash, set it to an empty value (it's a fake hash, hence the name of the method):

Reviewer: ```suggestion
Set the hash to an empty string:
```

- A built-in fetcher will download and cache files at evaluation time and produce a [store path](https://nixos.org/manual/nix/stable/glossary#gloss-store-path).
  A Nixpkgs fetcher will create a ([fixed-output](https://nixos.org/manual/nix/stable/glossary#gloss-fixed-output-derivation)) [derivation](https://nixos.org/manual/nix/stable/glossary#gloss-derivation), and files are downloaded at build time.
- Built-in fetchers will invalidate their cache after [`tarball-ttl`](https://nixos.org/manual/nix/stable/command-ref/conf-file#conf-tarball-ttl) expires, and will require network activity to check if the cache entry is up to date.
  Nixpkgs fetchers only re-download if the specified hash changes or the store object is not available.
- Built-in fetchers do not use [substituters](https://nixos.org/manual/nix/stable/command-ref/conf-file#conf-substituters).
  Derivations produced by Nixpkgs fetchers will use any configured binary cache transparently.

This significantly reduces the time needed to evaluate code in Nixpkgs, and allows [Hydra](https://nixos.org/hydra) to retain and re-distribute sources used by Nixpkgs in the [public binary cache](https://cache.nixos.org).

> 

Reviewer: ```suggestion
```

For these reasons, the Nix built-in fetchers are not allowed in Nixpkgs code.

> The following table summarizes the differences:

Reviewer: ```suggestion
The following table summarises the differences:
```

```
:::

## Obtaining hashes securely {#sec-pkgs-fetchers-secure-hashes}

Let's say Man-in-the-Middle (MITM) sits close to your network.
Then instead of fetching source you can fetch malware, and instead of the actual source hash, you get the hash of malware.
Here are security considerations for this scenario:

- `http://` URLs are not secure to prefetch hashes.

- Upstream hashes should be obtained via a secure protocol.

- `https://` URLs are secure when using `nix-prefetch-*` or for upstream hashes.

- `https://` URLs are secure relying on "trust on first use" *only if* you use one of the listed fake hashes.
  If you use any other hash, `fetchurl` will pass `--insecure` to `curl` and may then degrade to HTTP in case of TLS certificate expiration.

[]{#fetchurl}
## `fetchurl` {#sec-pkgs-fetchers-fetchurl}

`fetchurl` returns a derivation which downloads content from a given URL and stores the unaltered contents within the Nix store.

It uses {manpage}`curl(1)` internally, and allows its behaviour to be modified by specifying a few attributes in the argument to `fetchurl` (see the documentation for attributes `curlOpts`, `curlOptsList`, and `netrcPhase`).

> ### File naming {#sec-pkgs-fetchers-fetchurl-file-naming}

Reviewer: ```suggestion
```

Don't think this is needed. The section is not that long.

Author: Thanks! I split this PR into two, and I applied this suggestion in #300429 which is the PR with the fetchurl changes.

- Built-in fetchers do not use [substituters](https://nixos.org/manual/nix/stable/command-ref/conf-file#conf-substituters).
  Derivations produced by Nixpkgs fetchers will use any configured binary cache transparently.

This significantly reduces the time needed to evaluate code in Nixpkgs, and allows [Hydra](https://nixos.org/hydra) to retain and re-distribute sources used by Nixpkgs in the [public binary cache](https://cache.nixos.org).
For these reasons, the Nix built-in fetchers are not allowed in Nixpkgs code.

The following table summarises the differences:

| Fetchers | Download | Output | Cache | Re-download when |
|-|-|-|-|-|
| `builtins.fetch*` | evaluation time | store path | `/nix/store`, `~/.cache/nix` | `tarball-ttl` expires, cache miss in `~/.cache/nix`, output store object not in local store |
| `pkgs.fetch*` | build time | derivation | `/nix/store`, substituters | output store object not available |

> :::{.note}

Author: ```suggestion
:::{.tip}
```

`pkgs.fetchFrom*` helpers retrieve _snapshots_ of version-controlled sources, as opposed to the entire version history, which is more efficient.
:::

## Caveats {#chap-pkgs-fetchers-caveats}

> Because Nixpkgs fetchers are fixed-output derivations, an [output hash](https://nixos.org/manual/nix/stable/language/advanced-attributes#adv-attr-outputHash) has to be specified.

Author: ```suggestion
Because Nixpkgs fetchers are fixed-output derivations, an [output hash](https://nixos.org/manual/nix/stable/language/advanced-attributes#adv-attr-outputHash) has to be specified, usually indirectly through a `hash` attribute.
```

> This hash refers to the derivation output, not the remote source itself!

Author: ```suggestion
This hash refers to the derivation output, which can be different from the remote source itself!
```

This has the following implications that you should be aware of:

- Use Nix (or Nix-aware) tooling to produce the output hash.

- When changing any fetcher parameters, always update the output hash.
  Use one of the methods from [](#sec-pkgs-fetchers-updating-source-hashes).

>   Otherwise, existing store objects that match the output hash will be re-used rather than fetching new content.
> 
> A similar problem arises while testing changes to a fetcher's implementation.
> If the output of the derivation already exists in the Nix store, test failures can go undetected.
> The [`invalidateFetcherByDrvHash`](#tester-invalidateFetcherByDrvHash) function helps prevent reusing cached derivations.

Author: ```suggestion
  Otherwise, existing store objects that match the output hash will be re-used rather than fetching new content.

  :::{.note}
  A similar problem arises while testing changes to a fetcher's implementation.
  If the output of the derivation already exists in the Nix store, test failures can go undetected.
  The [`invalidateFetcherByDrvHash`](#tester-invalidateFetcherByDrvHash) function helps prevent reusing cached derivations.
  :::
```

## Updating source hashes {#sec-pkgs-fetchers-updating-source-hashes}

> There are several ways to obtain a remote source's hash.

Author: ```suggestion
There are several ways to obtain the hash corresponding to a remote source.
Unless you understand how the fetcher you're using calculates the hash from the downloaded contents, you should use [the fake hash method](#sec-pkgs-fetchers-updating-source-hashes-fakehash-method).
```

1. Prefetch the source with [`nix-prefetch-<type> <URL>`](https://search.nixos.org/packages?buckets={%22package_attr_set%22%3A[%22No%20package%20set%22]%2C%22package_license_set%22%3A[]%2C%22package_maintainers_set%22%3A[]%2C%22package_platforms%22%3A[]}&query=nix-prefetch), where `<type>` is one of

   - `url`
   - `git`
   - `hg`
   - `cvs`
   - `bzr`
   - `svn`

  The hash is printed to stdout.

2. Prefetch by package source (with `nix-prefetch-url '<nixpkgs>' -A <package>.src`, where `<package>` is package attribute name).
   The hash is printed to stdout.

   This works well when you've upgraded the existing package version and want to find out new hash, but is useless if the package can't be accessed by attribute or the package has multiple sources (`.srcs`, architecture-dependent sources, etc).

3. Upstream hash: use it when upstream provides `sha256` or `sha512`.
   Don't use it when upstream provides `md5`, compute `sha256` instead.

>    A little nuance is that `nix-prefetch-*` tools produce hashes encoded with the custom Nix `base32` variant, but upstream usually provides hexadecimal (`base16`) encoding.

Author: ```suggestion
   A little nuance is that `nix-prefetch-*` tools produce hashes with the `nix32` encoding (a Nix-specific base32 adaptation), but upstream usually provides hexadecimal (`base16`) encoding.
```

   Fetchers understand both formats.
   Nixpkgs does not standardise on any one format.

   You can convert between hash formats with [`nix-hash`](https://nixos.org/manual/nix/stable/command-ref/nix-hash).

4. Extract the hash from a local source archive with `sha256sum`.
   Use `nix-prefetch-url file:///path/to/archive` if you want the custom Nix `base32` hash.

> 5. Trust on first use: set the hash to one of

Author: Every other method is also trust-on-first-use depending on how you see it. This is the only method that someone can follow without having to learn beforehand how exactly a fetcher calculates the hash (e.g. `fetchzip` uses recursive hash, so blindly using `nix-prefetch-url` or `sha256sum` won't give the correct hash).

```suggestion
5. []{#sec-pkgs-fetchers-updating-source-hashes-fakehash-method} In your package recipe, set the hash to one of
```

   - `""`
   - `lib.fakeHash`
   - `lib.fakeSha256`
   - `lib.fakeSha512`

>    in the package recipe, attempt to build, and extract the correct hash from error messages.

Author: ```suggestion
   Once you attempt to build, error messages with the calculated hashes will show up, which you then replace in the recipe.
```

   :::{.warning}
   You must use one of these four fake hashes and not some arbitrarily-chosen hash.
   See [][#sec-pkgs-fetchers-secure-hashes] for details.
   :::

> 
>    This is last resort method when reconstructing source URL is non-trivial and `nix-prefetch-url -A` isn’t applicable.

Author: ```suggestion
```

:::{.example #ex-fetchers-update-fod-hash}

> # Update source hash with "trust on first use"

Author: ```suggestion
# Update source hash with the fake hash method
```

## Caveats {#chap-pkgs-fetchers-caveats}

Because Nixpkgs fetchers are [fixed-output derivations](https://nixos.org/manual/nix/stable/glossary#gloss-fixed-output-derivation), they need you to specify a hash (this is what turns them into a fixed-output derivation).
This is usually done through a `hash` attribute.

> However, **this hash is not a hash of the contents that are downloaded**, it is a hash that corresponds to the derivation output itself, and is calculated by Nix.

Author: Thank you! Indeed, after reading this section again I wasn't happy with how things were explained, and tried to improve it.

```shell
$ nix-build
(some output removed for clarity)
error: hash mismatch in fixed-output derivation '/nix/store/7yynn53jpc93l76z9zdjj4xdxgynawcw-version.drv':
         specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
            got:    sha256-BZqI7r0MNP29yGH5+yW2tjU9OOpOCEvwWKrWCv5CQ0I=
error: build of '/nix/store/bqdjcw5ij5ymfbm41dq230chk9hdhqff-version.drv' failed
```

A similar problem arises while testing changes to a fetcher's implementation.
If the output of the derivation already exists in the Nix store, test failures can go undetected.
The [`invalidateFetcherByDrvHash`](#tester-invalidateFetcherByDrvHash) function helps prevent reusing cached derivations.

[]{#fetchurl}
## `fetchurl` {#sec-pkgs-fetchers-fetchurl}

Downloads content from a given URL and stores the unaltered contents within the Nix store.

`fetchurl` uses {manpage}`curl(1)` internally, and allows its behaviour to be modified by specifying a few attributes in the argument to `fetchurl` (see the documentation for attributes `curlOpts`, `curlOptsList`, and `netrcPhase`).

### File naming {#sec-pkgs-fetchers-fetchurl-file-naming}

> The name of the file within the Nix store is controlled by the hash given to `fetchurl`, and also the `name` (or `pname` and `version`) values.

Reviewer 2: > The name of the file within the Nix store

I believe the official term is "derivation name"

Author: The Nix manual uses "derivation name" for the symbolic part of the name (which in the case of `fetchurl` would be determined by the `name` or `pname` + `version` attributes). I wanted to refer to the file name in this case, but the naming gets very confusing very fast (see below), so I prefer to stick to "name of the file within the Nix store" for now. At some point I want to attach a glossary to the Nixpkgs manual to make it easier to refer to names and their definitions, and at that point I hope I can start clarifying definitions in other manuals as well.

Why it gets very confusing very fast:
- In theory what I want to refer should be called the store path, but the [Nix manual glossary](https://nixos.org/manual/nix/stable/glossary#gloss-store-path) has a definition for store path which is not coherent with the [store path section](https://nixos.org/manual/nix/stable/store/store-path) in the same manual.
- Calling it a "file within the Nix store" is also not entirely true because there can be Nix stores that aren't part of the filesystem, but I couldn't find any better term that is easily understandable for the main audience of the manual.

Given that most people reading the manual will think of the Nix store as the one in their local filesystem, I chose to use that. The people who are aware of other store types and its quirks should be able to understand that the "file" part doesn't matter and this section is talking about the name of the object in the store more generally.

Reviewer 2: ah, here you are referring to hash+name, but i believe earlier you used the same wording to describe just the derivation name?

Author: It was later when describing the `name`, `pname`, and `version` attributes. I just pushed a change to be more specific on what kind of "name" I'm talking about, while trying as much as possible to avoid more confusion.

Reviewer: ```suggestion
The resulting store path is determined by the hash given to `fetchurl`, and also the `name` (or `pname` and `version`) values.
```

And maybe use the Nix manual link from the line below, then we don't need an entire extra sentence.

Author: I split this PR into two. The `fetchurl` changes are in #300429 where I applied this suggestion.

}
```

> **This will produce the same output as before!**.

Author: ```suggestion
**This will produce the same output as before!**.
To determine the new hash, set it to an empty value (it's a fake hash, hence the name of the method):
```

}
```

When building the package, use the error message to determine the correct hash:

```shell
$ nix-build

> ...

Author: I prefer to be verbose in this case: most people who are familiar with CLIs will understand both "..." and "(some output removed for clarity)", but there is a small amount of less experienced folks who will get a wall of text of output and think something went wrong and never actually read the output.

Using "(some output removed for clarity)" gives them indirect encouragement to read through the wall of text until they find the text they should be reading.

```suggestion
(some output removed for clarity)
```

Reviewer: Hm, I'd rather stick to the widely used convention and (at some point in the future) add a page on how to read examples.

Author: I decided to defer on this one since there is already a bunch of content (from previous PRs) where I used `(some output removed for clarity)` or `(output removed for clarity)`, there are also places in the manual where `...` is used, and other places where nothing is used.

I'd like to make this part of the linting tool for docs content, which is in [the roadmap](https://discourse.nixos.org/t/a-roadmap-for-the-documentation-ecosystem/42328) for docs, so let's make this decision in the issue for the linting tool itself. And then we can apply it to all existing content to make things consistent at once.

error: hash mismatch in fixed-output derivation '/nix/store/7yynn53jpc93l76z9zdjj4xdxgynawcw-version.drv':
         specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
            got:    sha256-BZqI7r0MNP29yGH5+yW2tjU9OOpOCEvwWKrWCv5CQ0I=
error: build of '/nix/store/bqdjcw5ij5ymfbm41dq230chk9hdhqff-version.drv' failed
```
:::

## Obtaining hashes securely {#sec-pkgs-fetchers-secure-hashes}

> Let's say Man-in-the-Middle (MITM) sits close to your network.
> Then instead of fetching source you can fetch malware, and instead of the actual source hash, you get the hash of malware.

Author: ```suggestion
It's always a good idea to avoid Man-in-the-Middle (MITM) attacks when downloading source contents.
Otherwise, you could unknowingly download malware instead of the intended source, and instead of the actual source hash, you'll end up using the hash of malware.
```

Here are security considerations for this scenario:

- `http://` URLs are not secure to prefetch hashes.

- Upstream hashes should be obtained via a secure protocol.

> - `https://` URLs are secure when using `nix-prefetch-*` or for upstream hashes.

Author: ```suggestion
- `https://` URLs give you more protections when using `nix-prefetch-*` or for upstream hashes.
```

> - `https://` URLs are secure relying on "trust on first use" *only if* you use one of the listed fake hashes.
>   If you use any other hash, `fetchurl` will pass `--insecure` to `curl` and may then degrade to HTTP in case of TLS certificate expiration.

Author: Mentioning `fetchurl` details will feel confusing here. If this is new information for anyone reading this, chances are very low that they will understand what "`fetchurl` will pass `--insecure` to `curl`" and what "TLS certificate expiration" means.

I'm adapting this to be most understandable to the audience which will find this section relevant.

```suggestion
- `https://` URLs are secure when using the [fake hash method](#sec-pkgs-fetchers-updating-source-hashes-fakehash-method) *only if* you use one of the listed fake hashes.
  If you use any other hash, the download will be exposed to MITM attacks even if you use HTTPS URLs.
```

Reviewer: No, security-related information should be concrete. This smudges away the facts. We can be a bit more clear and say

> will pass `--insecure` to the underlying call to` curl`

Author: Added this information, I think it covers all audiences now.
